File: php/check_booking/src/Model/Lesson/Status.php
<?php

namespace Egugue\BookingSerivce\CheckBooking\Model\Lesson;

use Egugue\BookingSerivce\CheckBooking\Util\Enum;
use InvalidArgumentException;

require_once __DIR__ . "/../../Util/Enum.php";

/**
 * A status a Lesson has.
 */
final class Status extends Enum
{
  /**
   * A status that a lesson has already reserved.
   */
  const RESERVED = "reserved";

  /**
   * A status that booking is available.
   */
  const AVAILABLE = "available";

  /**
   * A status that a Lesson was finished.
   */
  const FINISHED = "finished";

  /**
   * A status that a lesson was cancelled.
   */
  const CANCELLED = "cancelled";

  /**
   * A status that a lesson was not yet assigned.
   */
  const UNASSIGNED = "unassigned";

  public static function from(string $bookingText): Status
  {
    if ($bookingText === "") {
      return self::UNASSIGNED;
    } else if ($bookingText === "予約済") {
      return self::RESERVED;
    } else if ($bookingText === "予約可") {
      return self::AVAILABLE;
    } else if ($bookingText === "終了") {
      return self::FINISHED;
    }

    throw new InvalidArgumentException("Unexpected bookingText: {$bookingText}");
  }
}

File: scala/de_booker/app/com/egugue/debooker/domain/lesson/Status.scala
package com.egugue.debooker.domain.lesson

/**
  * A status that a lesson has
  */
sealed abstract class Status(val value: String)

object Status {

  case object Reserved extends Status("reserved")

  case object Finished extends Status("finished")

  case object Unassigned extends Status("unassigned")

  case object Cancelled extends Status("cancelled")

  case object Available extends Status("available")

  val values:Seq[Status] = Seq(Reserved, Finished, Unassigned, Cancelled, Available)

  /**
    * Retrieve a status of the given value.
    * If none of status has value, throw [[IllegalArgumentException]].
    */
  def of(value: String): Status = Status.values
    .find(status => status.value == value)
    .getOrElse(throw new IllegalArgumentException("Unexpected value: " + value))
}
